src/ai/ai.cpp

```diff
@@ -150,86 +377,52 @@ static void internal_get_alliance_targets(sys::state& state, dcon::nation_id n,
 void form_alliances(sys::state& state) {
 	static std::vector<dcon::nation_id> alliance_targets;
 	for(auto n : state.world.in_nation) {
-		if(!n.get_is_player_controlled() && n.get_ai_is_threatened() && !(n.get_overlord_as_subject().get_ruler())) {
-			alliance_targets.clear();
-			internal_get_alliance_targets(state, n, alliance_targets);
-			if(!alliance_targets.empty()) {
-				std::sort(alliance_targets.begin(), alliance_targets.end(), [&](dcon::nation_id a, dcon::nation_id b) {
-					if(estimate_strength(state, a) != estimate_strength(state, b))
-						return estimate_strength(state, a) > estimate_strength(state, b);
-					else
-						return a.index() > b.index();
-				});
-				nations::make_alliance(state, n, alliance_targets[0]);
+		if(n.get_is_player_controlled())
+			continue;
+		if((n.get_overlord_as_subject().get_ruler()))
+			continue;
+		if(state.world.nation_get_diplomatic_points(n) < state.defines.alliance_diplomatic_cost)
+			continue;
+		alliance_targets.clear();
+		internal_get_alliance_targets(state, n, alliance_targets);
+		if(!alliance_targets.empty()) {
+			std::sort(alliance_targets.begin(), alliance_targets.end(), [&](dcon::nation_id a, dcon::nation_id b) {
+				if(estimate_strength(state, a) != estimate_strength(state, b))
+					return estimate_strength(state, a) > estimate_strength(state, b);
+				else
+					return a.index() > b.index();
+			});
+			command::execute_ask_for_alliance(state, n, alliance_targets[0]);
```

Comment:
so, assuming that you have excluded human-controlled nations from the list of possibilities, the list of possibilities should only include valid targets that would accept the alliance. Thus you can go directly to `make_alliance` and skip the redundant checks. The reason we always exclude human-controlled nations from the list of potential targets is to not annoy the human player. The human player is assumed to proactively ask for alliances that he or she wants, and if the human player isn't asking, then they are assumed not to want one. Hence, we don't annoy them with asks that are expected to be rejected.